### PR TITLE
fix(compiler): properly bind to properties that don't have matching attr name

### LINKED ISSUE #619: Correct capitalization in bindings
- attr: `tabindex` vs prop:`tabIndex`
- attr: `inner-html` vs prop:`innerHTML`
others.

We should have a general list of overrides which correct these discrepancies in the compiler.

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Fixes #619

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
This is the first attempt at #619 but with few remark: I'm completely buying the idea of "fixing" property names where both attr _and_ property exists (ex. `tabindex` => `tabIndex`), but I'm less sure about the situation where there is no attribute but a property exists. Why do we want to "invent" attribute names? Shouldn't people simply bind to property names in this case? 

Currently the build is failing on the Dart version compilation, trying to figure out what is going on, as I'm a bit puzzled by this failure...

**Comment by vicb** [comment]
@pkozlowski-opensource Seems Travis fails because of `innerHtml` vs `innerHTML` (case)

**Comment by pkozlowski-opensource** [comment]
@vicb yeh. That's bad since in JS it is `innerHTML`.... OK, I will add special casing for inner html...

**Comment by pkozlowski-opensource** [comment]
So, the solutions I can see right now is to:
* have a separate / different mapping tables for Dart and JS (this makes me sad...) + facade for a getter
* special-case `inner-html`

But yeh, I really don't feel good about "inventing" the `inner-html` attribute name.

@mhevery could we discuss this during / after the today's standup?

**Comment by pkozlowski-opensource** [comment]
Remaining tasks:
* [ ] - decide what to do about innerHtml
* [ ] - find out what are the other attr => property mappings (ex.: `scrollLeft`, `scrollTop`)

**Comment by pkozlowski-opensource** [comment]
@mhevery I believe this is ready for the review. Just 2 items I'm unsure of:
* I had to [hack a bit for innerHtml](https://github.com/pkozlowski-opensource/angular/commit/28e37d2129d33b0b00314a8ecbb087fa5e3cafe7#diff-a7cac6bcdeec8aa9d185a9afd1c13685R185) due to #789 
* I was kind of surprised to see that what I'm getting in the `ElementBinderBuilder` compilation step were "un-normalized" attribute names - that is - for `<div [scroll-top]="expr">` I would get prop name being `scroll-top` while from your description I was assuming that it gets normalized to `scrollTop` in one of the previous steps. So the question is: should I expect `scroll-top` or `scrollTop` inside `ElementBinderBuilder` (currently it gets `scroll-top`)?

**Comment by benouat** [comment]
I don't really get the point to create _fake_ attributes that don't exist in the first place.

@pkozlowski-opensource The goal being to be able to bind directly to properties, why not using directly the property name without any conversion (I know it is not solving the different case issue between Dart and JS, but from a user point of view, you will write either Dart or JS) ? Maybe I am missing the point on that topic...

Regarding `inner-html` what would be the difference between:

```
<div [innerHTML]="myContent"></div>
```
and
```
<div>{{myContent}}</div>
```

**Comment by pkozlowski-opensource** [comment]
@benouat there are 2 things to observe here:
* if you type `<div innerHTML="foo">` a browser will convert it to `<div innerhtml="foo">` so we can't take advantage of mixed case in templates
* difference between `[inner-html]="foo"` and `{{foo}}` as the same as `element.innerHTML = foo` vs. `element.textContent  = foo`

**Comment by mhevery** [comment]
The conversion from dash-case to camelCase should happen in `ElementBinderBuilder`. Let's fix it there if it already is not fixed

File: modules/angular2/src/facade/dom.dart [hunk 1]
```diff
@@ -34,6 +35,11 @@ void gc() {
 
 final identitySanitizer = new IdentitySanitizer();
 
+var attrToPropMap = {
+  "tabindex": "tabIndex",
```
Comment (mhevery) [inline modules/angular2/src/facade/dom.dart]: readonly vs readOnly as well

File: modules/angular2/src/facade/dom.es6 [hunk 2]
```diff
@@ -14,6 +14,11 @@ export var gc = window.gc ? () => window.gc() : () => null;
 export var CssRule = window.CSSRule;
 export var CssKeyframesRule = window.CSSKeyframesRule;
 
+export var attrToPropMap = {
```
Comment (mhevery) [inline modules/angular2/src/facade/dom.es6]: can we have some system where we keep the common ones in one place and only keep the deltas in the facades?

**Comment by mhevery** [comment]
Please fix the comments and it should be ready to go.

**Pull request closed without merging**